config: synchronize `ALTER SYSTEM` parameters with LaunchDarkly

aalexandrov opened the pull request:
This implements the bulk of the work required to close #15230.

I'm currently working on migrating the `backend.rs` to the `SessionClient` based approach so so the sync task doesn't need to go through a `pgwire` connection (as discussed with @benesch offline). The rest of the code is more or less stable, so I'm open to review feedback.

### Motivation

  * This PR adds a known-desirable feature.

Closes #15956.

### Tips for reviewer

I've tried to split changes into digestible commits.

**Update**:
- The last two commits rework the implementation as suggested by @benesch.
  - I'm now using `SessionClient` for the backend in `backend.rs`.
  - I'm now using a direct dependency to `SystemVars` for the `SynchronizedParameters` in `params.rs`.
  - The `system_parameter_sync` is effectively unchanged compared to the original version.
- I had to move the setup code to `environmentd::serve` in order to get access to the `SystemVars` instance.

The starting point of the review should be:

- The setup code in the `environmentd::serve` method.
- The `system_parameter_sync` method in the new `mz_adapter::config` crate.
- After that, look at the `backend.rs`, `frontend.rs`, and `params.rs` in `mz_adapter::config`.

Regarding **changes of existing code**:
- ~I had to move the `Value` trait definition and some implementations from `mz_adapter::session::vars` to `mz_config::value`.~
- ~If we don't want to duplicate constants, I'll also have to do somethings similar with the `ServerVar<T>` declarations.~
- Alternatively, I can add all the code into a sub-folder of `src/adapter/src`. (done :heavy_check_mark:)

Regarding **testing**:
- So far I've tested locally, reproducing various failure scenarios, and the code seems to be working as expected.
- @philip-stoev created a stub for a test that I will extend tomorrow.

### TODOs

- [x] Move the forks of all patched libraries to `MaterializeInc`, open PRs for the patches, and refer to these PRs in `deny.toml`.
- [x] Fix conflicting dependency errors in the CI builds.
- [x] Fix the log messages so they adhere to [the recently adopted guidelines](https://github.com/MaterializeInc/materialize/commit/9423f03c44daee7aab4935802231530ae66d13f2).
- [x] Move all code to `adapter`.
- [x] Remove public methods that re not needed for now (`ensure_computed`, the `main` method that runs the sync task as a separate process).
- [x] Mark `config_has_synced_once` as a non-public feature (see [comment](https://github.com/MaterializeInc/materialize/pull/16089#discussion_r1034367971)).
- [x] Implement an alternative `SystemParameterBackend` that uses the `SessionClient` instead of a `tokio_postgres` connection.
- [ ] Add an end-to-end test (might be a separate PR).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

benesch reviewed (changes requested):
If we can't trust the LaunchDarkly Rust SDK to run in `environmentd`, my gut take is that we shouldn't use it. I would rather write our own integration against their API directly than attempt to isolate whatever shoddy code is in the Rust SDK in its own process. Is the code actually that bad? What's the source of our hesitation? Have we observed the code misbehaving, or are we just scared because of the warning that says the SDK is not production ready?

On a meta level, the question of whether to introduce a new process for a component discussed should be discussed in the design document for that feature. I would like to be very deliberate about where we introduce process boundaries on the database side of Materialize.

teskje commented:
> Is the code actually that bad? What's the source of our hesitation? Have we observed the code misbehaving, or are we just scared because of the warning that says the SDK is not production ready?

AFAIK (correct me if I'm wrong @aalexandrov) we are only scared of the warning. 

I think @benesch has a point that we probably don't need to take this level of precaution based on a warning alone. Most likely the warning doesn't exist because something is wrong with the code but because LD doesn't have applied their usual QA procedures to the Rust SDK, so they don't want to make promises. We should also not expect version updates to be backward-compatibly, but I think that's something we can handle.

I don't think we usually apply an extensive vetting process to new crates we start using, and I see no reason to distrust the quality of an LD crate more than any other open-source crate. The GitHub project for the Rust SDK doesn't show any open bugs, only feature requests. I don't see unsafe code, and there are a bunch of unit tests, which is a good sign.

The way I envision an environmentd integration to work is that there will be a new command line flag for passing the LD SDK key. If that flag is set, environmentd will start a background task that does the config syncing from LD. If it turns out that our worst fears are true and the Rust SDK crashes environmentd all the time, we can still disable that config task by adjusting the pod definition to not pass the LD flag to environmentd, while we work on a proper fix.

benesch commented:
> I think @benesch has a point that we probably don't need to take this level of precaution based on a warning alone. Most likely the warning doesn't exist because something is wrong with the code but because LD doesn't have applied their usual QA procedures to the Rust SDK, so they don't want to make promises. We should also not expect version updates to be backward-compatibly, but I think that's something we can handle.
> 
> I don't think we usually apply an extensive vetting process to new crates we start using, and I see no reason to distrust the quality of an LD crate more than any other open-source crate. The GitHub project for the Rust SDK doesn't show any open bugs, only feature requests. I don't see unsafe code, and there are a bunch of unit tests, which is a good sign.

Exactly! There are hundreds and hundreds of crates in our dependency set that are far more likely to have bugs than the LD crate.

aalexandrov commented:
@benesch: I [forked and updated](https://github.com/aalexandrov/) the `reqwest` dependency in the `rust-server-sdk`.

However, there is still the following packages that need to be upgraded:

- `cfg-if` (`v0.1.10 ⇒ v1.0.0`)
- `crossbeam-epoch` (`0.8.2 ⇒ 0.9.5`)
- `crossbeam-utils` (`0.7.2 ⇒ 0.8.7`)
- `hyper-rustls` (`0.22.1 ⇒ 0.23.1`)
- `lru` (`0.7.8 ⇒ 0.8.1`)
- `semver` (`0.10.0 ⇒ 1.0.14`)
- `time` (`0.1.43 ⇒ 0.3.17`)
- `uuid (`0.8.2 ⇒ 1.2.2`)
- `wasi` (`0.10.2+wasi-snapshot-preview1 ⇒ 0.11.0+wasi-snapshot-preview1`)

There are some more [transitive errors](https://buildkite.com/materialize/tests/builds/45814#0184bed8-6400-46c7-a292-06d82896af1a) that should be resolved if we fix the above.

Given the number of dependencies in the above list, should I just chisel at my LD SDK fork until we have them all fixed, or should we pivot to some of the other solutions (writing our own client, running the SDK in a separate process)?

benesch commented on src/adapter/src/catalog/config.rs:
```diff
@@ -56,6 +57,10 @@ pub struct Config<'a, S> {
+    /// A optional frontend used to pull system parameters for initial sync in
+    /// Catalog::open. A `None` value indicates that the initial sync should be
+    /// skipped.
+    pub system_parameter_frontend: Option<Arc<SystemParameterFrontend>>,
```
This seems like exactly the kind of coupling between components we were hoping to avoid? Gut take is that we should either avoid this coupling, or accept the coupling and use that to simplify the code. But having half of the code coupled and half of the code decoupled seems like we wind up with the downsides of both.

IMO, seems easiest to accept the coupling and move the code into the adapter directly. Wouldn't need to deal with `tokio_postgres` and the network failures, and you'd have access to the `ServerVar` defaults.

aalexandrov replied:
> This seems like exactly the kind of coupling between components we were hoping to avoid?

Unless we introduce some form hook that can be used to extend the catalog state initialization there is no way to avoid this coupling. We need the `config_has_synced_once` value to determine whether to override the state or not, so something like:

```rust
trait InitCatalogState {
    type E: Into<AdapterError>
    async fn state(state: &mut CatalogState) -> Result<(), E>;
}
```
and then pass an `InitCatalogState` type instead of the `system_parameter_frontend` in all `Config`.

However, it seems more pragmatic to just move everything under `adapter` for now...

aalexandrov replied:
> IMO, seems easiest to accept the coupling and move the code into the adapter directly

I did that in the last two commits of the updated PR.

benesch commented on src/adapter/src/session/vars.rs:
```diff
@@ -281,7 +280,15 @@ static ALLOWED_CLUSTER_REPLICA_SIZES: Lazy<ServerVar<Vec<String>>> = Lazy::new(|
+static CONFIG_HAS_SYNCED_ONCE: ServerVar<bool> = ServerVar {
+    name: UncasedStr::new("config_has_synced_once"),
+    value: &false,
+    description: "Boolean flag indicating that the remote configuration was synchronized at least once (Materialize).",
```
I don't think we want this becoming part of the stable interface of Materialize. Can we add a new `internal: bool` flag that hides parameters from `SHOW` and `SHOW ALL` unless you're logged in as `mz_system`?

aalexandrov replied:
Will do.

aalexandrov replied:
Done.

aalexandrov replied:
Done.

aalexandrov commented on src/config/src/bin/configd.rs:
```diff
@@ -0,0 +1,106 @@
+        eprintln!("computed: fatal: {:#}", err);
+        process::exit(1);
+    }
+}
```
Yes, my intention was to omit this code when I merge this PR, I should have created an explicit commit that deletes it when I was rebasing, sorry. I just want to have it preserve it in the history so it's easy to revert to that quickly if needed (and similarly with the pgwire-based backend).

aalexandrov replied:
Done.

aalexandrov commented:
@benesch the diff is getting quite large. Should I split out only the addition of the launchdarkly SDK to `mz-adapter` so we can merge that tomorrow? This will minimize the diff here and will allow me to spend less time resolving conflicts in `Cargo.lock` if for some reason we don't merge the rest of the code before next Monday.

benesch commented:
> This will minimize the diff here and will allow me to spend less time resolving conflicts in Cargo.lock if for some reason we don't merge the rest of the code before next Monday.

That sounds great!

aalexandrov commented on src/adapter/src/catalog/config.rs:
```diff
@@ -12,6 +12,7 @@ use std::net::Ipv4Addr;
 use std::num::NonZeroUsize;
 use std::sync::Arc;
 
+use mz_config::SystemParameterFrontend;
```
Fixed.

aalexandrov commented:
The first commit is #16395, I'll add the remaining agreed code changes to PR tomorrow.

aalexandrov commented on src/config/src/lib.rs:
```diff
@@ -0,0 +1,115 @@
+/// Start configd as a subprocess and ensure that it is restarted on failure
+/// until a termination signal is received.
+///
+/// Strictly speaking this is not needed for the current configuration.
```
I removed that.

aalexandrov commented on src/config/src/lib.rs:
```diff
@@ -0,0 +1,115 @@
+        if frontend.pull(&mut params) {
+            backend.push(&mut params).await;
+        }
+        time::sleep(sleep_duration).await;
```
Done.

aalexandrov commented:
@benesch I think I addressed all comments from the original review. Changes are mostly in the last two commits, as well as the commits that touch `SystemVars`. I can squash the former so we don't keep the interim implementation before merging.

teskje reviewed (approved):
LGTM. I have added a couple suggestions for improving documentation.

The `SessionClient` API seems very verbose to me, but I assume that is because that client is only used internally and there was no need to make it more ergonomic. I agree that changing this would be out-of-scope for this PR.

aalexandrov commented on src/adapter/src/session/vars.rs:
```diff
@@ -992,6 +992,46 @@ impl SystemVars {
     }
 
+    /// Checks if a configuration parameter named is currently
+    /// using its default.
```
Sorry, forgot to update the rustdoc string after changing the API of the method (originally it was not accepting a `value` parameter).

aalexandrov commented on src/adapter/src/config/params.rs:
```diff
@@ -0,0 +1,150 @@
+
+    /// Try to modify the in-memory `value` for `name` backing this
+    /// [SynchronizedParameters] instace, calling `SystemVars::set` iff `value`
+    /// is the default for this `name` and `SystemVars::reset` otherwise.
```
Yes.

aalexandrov commented on src/adapter/src/config/backend.rs:
```diff
@@ -7,10 +7,19 @@
 // by the Apache License, Version 2.0.
 
-use std::net::SocketAddr;
+#![allow(unused)]
```
Leftover from from the initial version of this file...

aalexandrov commented on src/adapter/src/catalog.rs:
```diff
@@ -854,12 +854,16 @@ impl CatalogState {
+        &mut self,
+        name: &str,
+        value: &str,
+    ) -> Result<bool, AdapterError> {
```
Done.

mjibson commented on src/adapter/src/catalog.rs:
```diff
@@ -2445,6 +2461,25 @@ impl<S: Append> Catalog<S> {
+                        )?;
+                    }
+
+                    catalog.state.set_config_has_synced_once()?;
```
I wasn't able to find a place where this is read and used other than this function. Did I miss something? If not, why do we need to set this?

aalexandrov replied:
Copy-pasting from the design doc:

> In addition to running that loop, we need to ensure that a one-time sync always runs synchronously as part of the initial `environmentd` creation. In order do that, we will introduce a dedicated  `SystemVar` boolean flag called `config_has_synced_once`. The `mz_environmentd::serve` call will block in cases where this value is `false` until we do the first synchronization. If LD is not reachable at this point in time, we will log an error and fallback to the defaults.

aalexandrov commented on src/adapter/src/catalog.rs:
```diff
@@ -2445,6 +2455,25 @@ impl<S: Append> Catalog<S> {
         for (name, value) in system_config {
             catalog.state.insert_system_configuration(&name, &value)?;
         }
+        if !catalog.state.system_config().config_has_synced_once() {
```
The default `CONFIG_HAS_SYNCED_ONCE` is `false`.

The idea is to run this block only once when the `environmentd` is initially created in order to assert that once the adapter is running it has synced once already. The block ends with a:
```rust
catalog.state.set_config_has_synced_once()?;
```
so it's not executed on restarts.

mjibson replied:
Oh right, I had forgotten that system vars are persisted to the stash. Makes sense now.

aalexandrov commented on src/adapter/src/config/params.rs:
```diff
@@ -0,0 +1,149 @@
+}
+
+impl ModifiedParameter {
+    pub fn as_alter_system(&self) -> String {
+        match self.is_default {
+            true => format!("ALTER SYSTEM RESET {}", self.name),
+            false => format!("ALTER SYSTEM SET {} = {}", self.name, self.value),
+        }
+    }
```
> This should return a statement directly.

I'm not sure things will work out fine with the proposed `as_stmt` for the following query:

```sql
ALTER SYSTEM SET allowed_cluster_replica_sizes = 'xlarge', '2xlarge';
```

The problem is that the `self.value: String` in this case will be `'xlarge', '2xlarge'` and if we pack it a string literal the resulting `sequence_alter_system_set` call will be confused. 

The general problem maybe is that `impl Var for T` provides primarily `T ⇔ String` conversions, whereas it might be better to focus on `T ⇔ Statement<Raw>` conversions and have default implementations `T ⇔ String` that go through `parse()` and `to_string()` of the resulting `Statement`.

I'll try to fix this in #16513 next. Once we should be able to use the extended `Var` API to directly construct the statement as you propose.

aalexandrov replied:
This actually works (I think), so I accepted the suggestion. I will revert if I encounter errors when adding the CI tests.

aalexandrov commented on src/adapter/src/config/backend.rs:
```diff
@@ -0,0 +1,169 @@
+                    tracing::debug!("set system parameter {} = {}", param.name, param.value)
+                }
+                Err(error) => {
+                    tracing::warn!("cannot execute `ALTER SYSTEM` query: {}", error)
```
Done.

aalexandrov commented on src/adapter/src/config/backend.rs:
```diff
@@ -0,0 +1,169 @@
+    }
+}
+
+/// Parse a single valid statement.
+fn parse_valid_stmt<T: Borrow<str>>(sql: T) -> Statement<Raw> {
+    parse(sql.borrow()).unwrap().pop().unwrap()
+}
+
```
See [my comment above](https://github.com/MaterializeInc/materialize/pull/16089#discussion_r1042601913).

I'll give it a try and try to reproduce a specific instance of the problems I can see with this at the moment...

aalexandrov commented on src/adapter/src/config/frontend.rs:
```diff
@@ -0,0 +1,95 @@
+
+// This is needed because some of the encosing `Config` structs that hold a
+// [SystemParameterFrontend] derive `Debug`.
+impl std::fmt::Debug for SystemParameterFrontend {
```
Nice!

aalexandrov commented on src/adapter/src/config/params.rs:
```diff
@@ -0,0 +1,149 @@
+    modified: BTreeSet<&'static str>,
+}
+
+impl Default for SynchronizedParameters {
```
The `system_parameter_sync` initializes a private `SynchronizedParameters` using the `default()` method for the main loop:

```rust
let mut params = SynchronizedParameters::default();
loop {
    backend.pull(&mut params).await;
    if frontend.pull(&mut params) {
        backend.push(&mut params).await;
    }
    interval.tick().await;
}
```

aalexandrov commented on src/environmentd/tests/util.rs:
```diff
@@ -252,6 +252,8 @@ pub fn start_server(config: Config) -> Result<Server, anyhow::Error> {
         egress_ips: vec![],
         aws_account_id: None,
+        launchdarkly_sdk_key: None,
+        config_sync_loop_interval: Duration::ZERO,
```
> The ZERO here won't cause a busy loop because the sdk key is None, and so doesn't ever start the ticker?

Yes, look for the following comments in the diff for the places in `environmentd::serve` that assert this:
```rust
// Initialize the system parameter frontend if `launchdarkly_sdk_key` is set.
// ...

// If system_parameter_frontend is present, start the system_parameter_sync loop.
// ...
```

aalexandrov commented on src/adapter/src/coord/sequencer.rs:
```diff
@@ -1929,8 +1931,13 @@ impl<S: Append + 'static> Coordinator<S> {
-        let row = Row::pack_slice(&[Datum::String(&variable.value())]);
-        Ok(send_immediate_rows(vec![row]))
+
+        if !variable.internal() || session.user() == SYSTEM_USER.deref() {
```
I can be a bit more concise and just change the signature of `internal()` to include `user: &User`.

aalexandrov commented on src/environmentd/src/lib.rs:
```diff
@@ -360,11 +399,24 @@ pub async fn serve(config: Config) -> Result<Server, anyhow::Error> {
+    // If system_parameter_frontend is present, start the system_parameter_sync loop.
+    if let Some(system_parameter_frontend) = system_parameter_frontend {
+        let system_parameter_backend = SystemParameterBackend::new(adapter_client).await?;
+        task::spawn(
```
I wonder if there is a way to ensure that we don't kill `environmentd` if the `system_parameter_sync` loop throws a panic. I think logging a `Sentry` error is probably and letting only the sync task die is the more robust solution here.

benesch replied:
As @doy-materialize mentioned on Slack, `mz_ore::panic::catch_unwind` is the way! There's an async variant too.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=16089) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MaterializeInc/materialize?pullRequest=16089) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MaterializeInc/materialize?pullRequest=16089) it.</sub>

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -2445,6 +2461,25 @@ impl<S: Append> Catalog<S> {
         for (name, value) in system_config {
             catalog.state.insert_system_configuration(&name, &value)?;
         }
+        if !catalog.state.system_config().config_has_synced_once() {
```
Let's make this way spammier! An info log at the start and end of config syncing would be great, and also an info log if we _don't_ need to do the initial config sync. Otherwise if LD goes down we're going to be mighty confused as to why some envd somewhere is wedged on boot, because we won't remember this code exists in the heat of that moment!

I think a debug log for each discovered system parameter would also be a good idea.

aalexandrov replied:
Done. I used only `--log-filter mz_adapter::catalog=debug,mz_adapter::config=debug` to see only the relevant output when  testing.

benesch commented on src/adapter/src/config/frontend.rs:
```diff
@@ -0,0 +1,88 @@
+    /// Construct a new [SystemParameterFrontend] instance.
+    pub fn new(ld_sdk_key: &str, ld_user_key: &str) -> Result<Self, anyhow::Error> {
+        let ld_config = ld::ConfigBuilder::new(ld_sdk_key).build();
+        let ld_client = ld::Client::build(ld_config).map_err(|e| anyhow::format_err!(e))?;
```
Confused about this `map_err`. Should be convertible to `anyhow::Error` without trouble?

```suggestion
        let ld_client = ld::Client::build(ld_config)?;
```

benesch commented on src/adapter/src/catalog.rs:
```diff
@@ -2445,6 +2461,25 @@ impl<S: Append> Catalog<S> {
+                let system_config = catalog.state.system_config();
+                let mut params = SynchronizedParameters::new(system_config.clone());
+
+                if system_parameter_frontend.start_and_initialize().await {
```
Wait, this doesn't seem right? `start_and_initialize` looks like it can fail? But don't we want to use `ensure_initialized` here instead, to actually wait for successful initialization?

aalexandrov replied:
> Wait, this doesn't seem right?

I think I wanted to ensure that we don't block new environments indefinitely if for some reason LD is down. I'll use `ensure_initialized` and we can see if this is a problem in practice.

benesch commented on src/environmentd/src/lib.rs:
```diff
@@ -286,6 +292,38 @@ pub async fn serve(config: Config) -> Result<Server, anyhow::Error> {
+                    })
+                    .transpose()
+                    .unwrap_or_else(|e| {
+                        tracing::error!("Cannot initialize system parameter frontend: {:?}", e);
```
This is surprising to me. AFAICT `SystemParameterFrontend::new` just builds some structs, and any errors are usage errors that we can just return with `?`, rather than via a `tracing::error!`. What am I missing?

aalexandrov replied:
Yes, this is a stupid leftover from a previous iteration when the constructor was also initializing the LD client.

benesch commented on src/environmentd/src/lib.rs:
```diff
@@ -286,6 +292,38 @@ pub async fn serve(config: Config) -> Result<Server, anyhow::Error> {
+                    .map(Arc::new)
+            },
+        )
+        .await
+        .map_err(|e| anyhow::anyhow!(e))?
```
```suggestion
        .await?
```

benesch commented on src/adapter/src/config/backend.rs:
```diff
@@ -0,0 +1,163 @@
+
+        self.prepare(stmt).await?;
+        let result = match self.session_client.execute(EMPTY_PORTAL.into()).await? {
+            ExecuteResponse::AlteredSystemConfiguraion => Ok(()),
```
I know it's pre-existing, but mind fixing the typo in "Configuraion"?

benesch commented:
> The SessionClient API seems very verbose to me, but I assume that is because that client is only used internally and there was no need to make it more ergonomic. 

Yep, exactly! I added this helper method recently which makes things a lot more ergonomic in telemetry.rs, though not directly usable here:

https://github.com/MaterializeInc/materialize/blob/54ada90bdb7e8406e56f62c1ece85f2c3870076e/src/adapter/src/client.rs#L99-L127

I'll try to find some time after this merges to generalize `introspection_execute_one` so that it is usable by the config backend.

benesch reviewed (commented):
Looking solid. Mostly cosmetic comments, although one big behavioral question about the first config sync.

How do we test this? I saw you mentioned the E2E test might wait for another PR. Is that still the plan? Seems safe enough to merge this as is, as it's all opt-in.

aalexandrov commented on src/adapter/src/config/mod.rs:
```diff
@@ -6,3 +6,46 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
+
+mod backend;
+mod frontend;
+mod params;
+
+use std::{sync::Arc, time::Duration};
+
+use tokio::time;
```
I've frequently made in organizing the code. I realize that at the moment there is no tool that reports them as lints, but is the the convention that we try to enforce in practice described somewhere?

aalexandrov commented:
> Looking solid. Mostly cosmetic comments, although one big behavioral question about the first config sync.
>
> How do we test this? I saw you mentioned the E2E test might wait for another PR. Is that still the plan? Seems safe enough to merge this without the test (modulo addressing comments), as it's all opt-in.

I have the tests almost ready but I prefer to open a separate PR for this (should be done by EOD today). And E2E test is still the plan, mostly because mocking the API endpoint seems too much overhead for now. If the test is flaky we can invest time to do that.

There was a bug in the first config sync as the catalog state changes were not persisted, so consequently the sync was running on every restart (I think this is fixed in the current revision, you might want to take another look at the `Catalog::open` changes).

To test both the boot-time and the periodic sync parts of the code, I suggest one of the following options:

1. Interpret `config_sync_loop_interval = 0s` as "I don't want to run the sync" (this will be my preferred option).
2. Move the `interval.tick().await;` at the beginning of the loop in `system_parameter_sync`.

In both cases I can construct a scenario (either by passing `0s` or some ridiculous large number, say `60m`), where `environmentd` starts and then I check for some parameter that I know has non-default value in LD.

I will then restart `environmentd` to test `system_parameter_sync`.

mjibson commented:
You could use an `Option<Duration>` instead of a sentinel value like 0s to mean "turn this off".

aalexandrov commented on src/environmentd/src/lib.rs:
```diff
@@ -360,11 +390,25 @@ pub async fn serve(config: Config) -> Result<Server, anyhow::Error> {
+                system_parameter_backend,
+                config.config_sync_loop_interval,
+            ))
+            .ore_catch_unwind(),
```
I tried appending a `map` to report any panics to Sentry, like that:

```rust
            .ore_catch_unwind()
            .map(|result| {
                if let Err(error) = result {
                    if let Ok(error) = error.downcast::<&str>() {
                        tracing::error!("unexpected {}", error);
                    } else {
                        tracing::error!("unexpected error");
                    }
                }
            }),
```

but I didn't see anything after hard-coding a panic in `system_parameter_sync`.

benesch replied:
And you tried this with the `--sentry-dsn` argument, or whatever it is, that you need to actually report to Sentry?

aalexandrov replied:
No, I meant that if I do this the error in stdout/stderr (whereas I see `tracing` messages from the main loop).

aalexandrov commented:
> You could use an `Option<Duration>` instead of a sentinel value like 0s to mean "turn this off".

I went with that option in #16536.
